src/sink_http.c

```diff
@@ -156,76 +178,246 @@ static int add_metrics(void* data,
+    } else {
+        /* There is partition data, separate name and partition */
+        partition = pos + 1;
+        name_len = pos - metric;
+    }
 
+    /* Early out if the metric name is empty */
+    if (name_len == 0) {
+        return 0;
+    }
+
+    /* Get prefix from config */
+    char* prefix = NULL;
+    uint16_t pre_len = 0;
+    if (httpconfig->use_prefix) {
+        prefix = config->prefixes_final[type];
+        pre_len = strlen(prefix);
+    }
+
+    /*
+     * Find the json object mapped to the partition string.
+     * Create one if there isn't one yet.
+     */
+    json_t* obj = NULL;
+    if (hashmap_get(json_map, partition, (void**)&obj) == -1) {
+        /* Create a new json object if it doesn't exist */
+        obj = json_object();
+        hashmap_put(json_map, partition, obj);
+    }
+
+    assert(obj != NULL);
+    _add_metric_to_json(obj, type, prefix, pre_len, metric, name_len, value, config);
     return 0;
 }
 
-static int json_cb(const char* buf, size_t size, void* d) {
+/*
+ * Callback function for dumping json object to string.
+ * @arg buf The data buffer to dump.
+ * @arg size Number of bytes to dump.
+ * @arg buf The output buffer.
+ * @return 0 on success.
+ */
+static int _json_cb(const char* buf, size_t size, void* d) {
     strbuf* sbuf = (strbuf*)d;
     strbuf_cat(sbuf, buf, size);
     return 0;
 }
 
-static int serialize_metrics(struct http_sink* sink, metrics* m, void* data) {
-    json_t* jobject = json_object();
-    const sink_config_http* httpconfig = (const sink_config_http*)sink->sink.sink_config;
+/*
+ * Escape a string and add it to the post buffer.
+ * @arg post_buf The post buffer.
+ * @arg value The string to escape and append.
+ * @arg len The length of the string to append. 0 if the full string should be appended.
+ * @arg curl The CURL object for string escaping.
+ */
+static void _escape_and_append_to_buffer(strbuf* post_buf,
+                                         const char* value,
+                                         int len,
+                                         CURL* curl) {
+    char* escaped_data = curl_easy_escape(curl, value, len);
+    strbuf_cat(post_buf, escaped_data, strlen(escaped_data));
+    curl_free(escaped_data);
+}
 
-    struct cb_info info = {
-        .jobject = jobject,
-        .config = sink->sink.global_config,
-        .httpconfig = httpconfig
-    };
-    /* produce a metrics json object */
-    metrics_iter(m, &info, add_metrics);
+/*
+ * Verify if a key exists in a query string or not. This function simply scans
+ * through the query string, hence it shouldn't be called often.
+ * @arg key The key to search. It should not contain characters '&' and '='.
+ * @arg partition_string The partition string to search in.
+ * @return True if the key exists as a key name(as key in key=123 but not in key123=123)
+ *         in the query string, false otherwise.
+ */
+bool _key_exists_in_query_string(const char* key,
```

Comment:
the two strings will be used in the final query string, so yeah, they are supposed to be whitespace free.

the function returns false when the key is an empty string and the query string is not, but it returns true if both strings are empty in this implementation. Adding another check in the function and unit test to cover both cases. Thanks for pointing it out.

And key should not be empty. I will add a check in the calling code.